numba/core/dispatcher.py

```diff
@@ -686,7 +686,10 @@ def typeof_pyval(self, val):
         try:
             tp = typeof(val, Purpose.argument)
         except ValueError:
-            tp = types.pyobject
+            if hasattr(val, '__array__'):
```

Comment:
I think that this should perhaps go into the "fingerprinting" code as part of the if-else tree here: https://github.com/numba/numba/blob/eb6b9790a803e54c0285f5328a5234aab8c830f4/numba/_typeof.cpp#L262
. This python Dispatcher code is a fallback path, it has no caching and requires Python C-API calls to reach it so it is quite slow. The reason I think it could work is that the function will need compiling for an Array type regardless of whether the object or the `object.__array__` is an array, and then the boxing will just handle the details of unpacking it?